Add Smart_Battery_Info MAVLink message

hendjoshsr71 opened the pull request:
This adds the MAVLink message 370 SMART_BATTERY_INFO to send serial number, cycles, capacity, product name, etc. 

PR to add the message to ArduPilot/mavlink here: https://github.com/ArduPilot/mavlink/pull/181
One small documentation change on the MAVLink side here: https://github.com/mavlink/mavlink/pull/1551

serial_number: I wasn't sure the best way/place to deal with the conversion from int32_t  to char[16].
``AP_BattMonitor.h: get_serial_number()``  that is used by ``AP_BattMonitor_UAVCAN.h`` not sure how to combine this together and limit the extra functions

``SIM_BattMonitor_SMBus.cpp`` Issue: The method to fill the registers for manufacturer_name and device_name causes the data from device name to write over all but the first char of the manufacturer_name. 

amilcarlucas commented on libraries/GCS_MAVLink/GCS_Common.cpp:
```diff
@@ -4465,6 +4468,55 @@ void GCS_MAVLINK::send_set_position_target_global_int(uint8_t target_system, uin
+
+    mavlink_msg_smart_battery_info_send(
+        chan,
+        instance, 						// Battery ID
```
Indent the comments with spaces

hendjoshsr71 replied:
Fixed

amilcarlucas commented:
The message is marked as "work-in-progress". Is it a good idea to bring it in? I do like your changes though.

peterbarker commented on libraries/AP_BattMonitor/AP_BattMonitor.cpp:
```diff
@@ -505,6 +505,31 @@ bool AP_BattMonitor::get_cycle_count(uint8_t instance, uint16_t &cycles) const
+        return false;
+    }
+
+    char device_name[SMBUS_READ_BLOCK_MAXIMUM_TRANSFER+1]{}; 
+    char manufacturer_name[SMBUS_READ_BLOCK_MAXIMUM_TRANSFER+1]{};
+
+    bool got_manufacturer_name = drivers[instance]->get_manufacturer_name(manufacturer_name);
+    bool got_device_name = drivers[instance]->get_device_name(device_name);
+
+    snprintf(product_name, buflen, "%s_%s", device_name, manufacturer_name);
+
+    return got_manufacturer_name && got_device_name;
```
```suggestion

    char manufacturer_name[SMBUS_READ_BLOCK_MAXIMUM_TRANSFER+1];
    if (!drivers[instance]->get_manufacturer_name(manufacturer_name)) {
        return false;
    }

    char device_name[SMBUS_READ_BLOCK_MAXIMUM_TRANSFER+1]; 
    if (!drivers[instance]->get_device_name(device_name)) {
        return false;
    }

    snprintf(product_name, buflen, "%s_%s", device_name, manufacturer_name);

    return true;
```

Avoids having to initialise the buffers on the stack.  Avoids unnecessary work if we fail to get one of the values.

Should we handle the case we get one or the other but not both, 'though?

hendjoshsr71 replied:
My thought would be to require both until a case comes up ... extra logic for what one would hope is a nonexistent case... (famous last words)

Or only require device_name as it is gives more info. than manufacturer name alone

hendjoshsr71 replied:
I made it such that it returns true if either both are available or if one is available but not both

peterbarker commented on libraries/AP_BattMonitor/AP_BattMonitor_Backend.h:
```diff
@@ -49,6 +49,15 @@ class AP_BattMonitor_Backend
     virtual bool get_cycle_count(uint16_t &cycles) const { return false; }
 
+    // return true if serial number can be provided and fills in serial number 
+    virtual bool get_serial_number(char *serial_number, uint8_t &serial_name_len) const { return false; };
```
Trailing semi-colons here are a bit messy.  Elsewhere too.

Don't want to pass in a reference for the length.

```suggestion
    virtual bool get_serial_number(char *serial_number, uint8_t serial_name_len) const { return false; }
```

similarly elsewhere

peterbarker commented on libraries/AP_BattMonitor/AP_BattMonitor_Backend.h:
```diff
@@ -49,6 +49,15 @@ class AP_BattMonitor_Backend
+    // return true if serial number can be provided and fills in serial number 
+    virtual bool get_serial_number(char *serial_number, uint8_t &serial_name_len) const { return false; };
+
+    // return true if device name can be provided and fills in product name
```
```suggestion
    // return true if device name can be provided and fills in device name
```

peterbarker commented on libraries/AP_BattMonitor/AP_BattMonitor_SMBus.cpp:
```diff
@@ -30,6 +30,74 @@ bool AP_BattMonitor_SMBus::get_cycle_count(uint16_t &cycles) const
+bool AP_BattMonitor_SMBus::get_serial_number(char *serial_number, uint8_t &serial_name_len) const {
+
+    if (_serial_number != -1) {
+    	snprintf(serial_number, serial_name_len, "%i", (signed) _serial_number);
```
Indenting failure.
```suggestion
        snprintf(serial_number, serial_name_len, "%i", (signed) _serial_number);
```

Please be careful not to mix spaces and tabs!

peterbarker commented on libraries/AP_BattMonitor/AP_BattMonitor_SMBus.cpp:
```diff
@@ -30,6 +30,74 @@ bool AP_BattMonitor_SMBus::get_cycle_count(uint16_t &cycles) const
+bool AP_BattMonitor_SMBus::get_serial_number(char *serial_number, uint8_t &serial_name_len) const {
+
+    if (_serial_number != -1) {
+    	snprintf(serial_number, serial_name_len, "%i", (signed) _serial_number);
```
Signed?  Really?
```suggestion
    	snprintf(serial_number, serial_name_len, "%u", (unsigned) _serial_number);
```

hendjoshsr71 replied:
_serial_number is an int32_t : What is the best way to deal with the cast? I'm rusty on such things.

peterbarker commented on libraries/AP_BattMonitor/AP_BattMonitor_SMBus.cpp:
```diff
@@ -30,6 +30,74 @@ bool AP_BattMonitor_SMBus::get_cycle_count(uint16_t &cycles) const
+{
+    if (_device_name_len != 0) {
+        return true;
+    } else {
```
Eliminate this else.  Saves a level of indenting for the rest of the function.

Similarly in the other methods below.

peterbarker commented on libraries/AP_BattMonitor/AP_BattMonitor_SMBus.cpp:
```diff
@@ -30,6 +30,74 @@ bool AP_BattMonitor_SMBus::get_cycle_count(uint16_t &cycles) const
+    } else {
+        _device_name_len = read_block(BATTMONITOR_SMBUS_DEVICE_NAME, _device_name, true) + 1;
+
+        if (_device_name_len > 0)
+        {
+            return true;
+        }
+    }
+
+    return false;
```
```suggestion
    return _device_name_len > 0;
```

Similarly in the like methods below

peterbarker commented on libraries/AP_BattMonitor/AP_BattMonitor_SMBus.cpp:
```diff
@@ -30,6 +30,74 @@ bool AP_BattMonitor_SMBus::get_cycle_count(uint16_t &cycles) const
+}
+
+// return true if device name can be provided and fills in product name
+bool AP_BattMonitor_SMBus::get_device_name(char * device_name) const
```
Can't do this without pass in the length of the `device_name` buffer.  Otherwise we'll end up with a stack overwrite on some product, and they *suck*,

Similarly in the other methods.

hendjoshsr71 replied:
fixed

peterbarker commented on libraries/AP_BattMonitor/AP_BattMonitor_SMBus.cpp:
```diff
@@ -30,6 +30,74 @@ bool AP_BattMonitor_SMBus::get_cycle_count(uint16_t &cycles) const
+    // already have manufacturer name return
+    if (_manufacturer_name_len != 0) {
+        return true;
+    } else {
```
Eliminate this else.  Saves a level of indenting.  Similarly elsewhere.

peterbarker commented on libraries/AP_BattMonitor/AP_BattMonitor_SMBus.cpp:
```diff
@@ -147,6 +215,50 @@ bool AP_BattMonitor_SMBus::read_word(uint8_t reg, uint16_t& data) const
     return true;
 }
 
+// read_block - returns number of characters read if successful, zero if unsuccessful
```
Can you say that this method was moved up unmodified?

hendjoshsr71 replied:
I needed to use the method for the device and manufacturer functions. So either move the method up or the functions down. Wasn't sure the best way to deal with it? 

There are 3-4 versions of the `AP_BattMonitor_SMBus::read_block` floating around in some of the SMBus derived classes. They are analogs that could likely just use this method instead.

peterbarker replied:
@hendjoshsr71 sorry, I wasn't clear in my original question.

I wasn't keen on making sure this was just byte-for-byte moved without modification from the subclass and was seeking assurance that it was just moved.

hendjoshsr71 replied:
Seems like I forgot to answer this one. Yes it was just a copy & paste. I'm not sure why the git changes for this function look so weird where I removed it.

peterbarker commented on libraries/AP_BattMonitor/AP_BattMonitor_SMBus.h:
```diff
@@ -50,6 +53,15 @@ class AP_BattMonitor_SMBus : public AP_BattMonitor_Backend
     // return true if cycle count can be provided and fills in cycles argument
     bool get_cycle_count(uint16_t &cycles) const override;
 
+    // return true if serial number can be provided and fills in cycles argument
```
```suggestion
    // return true if serial number can be provided and fills in serial number argument
```

So close ;-)

peterbarker commented on libraries/GCS_MAVLink/GCS_Common.cpp:
```diff
@@ -3492,6 +3492,9 @@ void GCS_MAVLINK::send_banner()
         }
     }
+
+    // send smart_battery_info MAVLink message for each monitor 
+    send_smart_battery_info();
```
There might be some push-back on this.

hendjoshsr71 replied:
Is there a better area/way to emit the message once at start-up or upon connection to GCS?  Still learning the code.

peterbarker replied:
Well, a better way might be for the GCS to probe for the relevant mavlink message and present it as appropriate.

ATM no GCS will handle the smart-battery-info message so....

hendjoshsr71 replied:
I added the smart_batery_info message to where it can be requested

peterbarker commented on libraries/GCS_MAVLink/GCS_Common.cpp:
```diff
@@ -4465,6 +4468,55 @@ void GCS_MAVLINK::send_set_position_target_global_int(uint8_t target_system, uin
 
+void GCS_MAVLINK::send_smart_battery_info(const uint8_t instance) const
+{
+    const AP_BattMonitor &battery = AP::battery();
```
Indenting is stuffed in here.  tabs vs spaces?

peterbarker commented on libraries/GCS_MAVLink/GCS_Common.cpp:
```diff
@@ -4465,6 +4468,55 @@ void GCS_MAVLINK::send_set_position_target_global_int(uint8_t target_system, uin
+    char serial_number[16]{};
+    battery.get_serial_number(instance, serial_number, ARRAY_SIZE(serial_number)); 
+
+    char product_name[50]{}; 
```
constants could be eliminated by using mavlink defines.
```suggestion
    char product_name[MAVLINK_MSG_SMART_BATTERY_INFO_FIELD_PRODUCT_NAME_LEN]{}; 
```

hendjoshsr71 replied:
Cool! Didn't know about those ones.

peterbarker commented on libraries/GCS_MAVLink/GCS_Common.cpp:
```diff
@@ -4465,6 +4468,55 @@ void GCS_MAVLINK::send_set_position_target_global_int(uint8_t target_system, uin
+{
+    const AP_BattMonitor &battery = AP::battery();
+    
+    for(uint8_t i = 0; i < battery.num_instances(); i++) {
+        if (battery.get_type(i) != AP_BattMonitor::Type::NONE) {
+            CHECK_PAYLOAD_SIZE(SMART_BATTERY_INFO);
+            send_smart_battery_info(i);
+        }
```
@WickedShell has grown to dislike this pattern, preferring to keep some state around indicating the last instance we successfully sent.

hendjoshsr71 replied:
The pattern for the other battery message like you describe has to be modified with the comment below due to where I have the message being sent from. Suggestions?

    for(uint8_t i = 0; i < battery.num_instances(); i++) {
        const uint8_t battery_id = (last_smart_battery_info_idx + 1) % AP_BATT_MONITOR_MAX_INSTANCES;
        if (battery.get_type(battery_id) != AP_BattMonitor::Type::NONE) {
            CHECK_PAYLOAD_SIZE(SMART_BATTERY_INFO);
            send_smart_battery_info(battery_id);
            last_smart_battery_info_idx = battery_id;
            //return true;
        } else {
            last_smart_battery_info_idx = battery_id;
        }
    }
    return true;

peterbarker reviewed (changes requested):
I love the addition to the simulated battery.

There might be some push-back on having the method returning a string for the serial number.  We'll see.

@magicrub you might be interested in options to pass more CAN battery information back to the GCS with this message.

This looks pretty good, really.

peterbarker commented on libraries/AP_BattMonitor/AP_BattMonitor.cpp:
```diff
@@ -571,6 +600,68 @@ bool AP_BattMonitor::reset_remaining(uint16_t battery_mask, float percentage)
+    // @Field: cyc: charge/discharge cycle count. UINT16_MAX: field not provided.
+    // @Field: ser: serial number in ASCII characters, 0 terminated. All 0: field not provided.
+    // @Field: name: product name encoded as 'device names'_'manufacturer'
+    // @Field: w: [newtons] battery weight. 0: field not provided. 
```
The unit is fiven elsewhere....

peterbarker commented on libraries/AP_BattMonitor/AP_BattMonitor.cpp:
```diff
@@ -571,6 +600,68 @@ bool AP_BattMonitor::reset_remaining(uint16_t battery_mask, float percentage)
+                        serial_number,
+                        product_name,
+                        0, // Newtons = 9.81*1E-3*g                 
+                        UINT16_MAX,
+                        UINT16_MAX,
+                        UINT16_MAX);
```
This a low-bandwidth message.  We could just use floats here and use a proper flag value - `NaN` - to say we don't know.

hendjoshsr71 replied:
Fixed

peterbarker reviewed (changes requested):
I'm going to mark this for discussion at the European DevCall.  I want more people to chime in on the send-as-part-of-banner thing.

peterbarker commented:
@hendjoshsr71 let me know if you'd like me to rebase this on the changes I've just made to the simulated-smbus stuff - it obviously conflicted.  I *think* the commit in your list here can just be removed.

hendjoshsr71 commented:
I got it. Thanks for the update! Adding in capacity and other fields to the sim and will reupload once working.

hendjoshsr71 commented:
Added design capacity and few more items to SITL_SMBus model

amilcarlucas commented:
This needs a rebase

hendjoshsr71 commented:
Rebase complete.

peterbarker commented on libraries/AP_BattMonitor/AP_BattMonitor.cpp:
```diff
@@ -512,6 +521,35 @@ bool AP_BattMonitor::get_cycle_count(uint8_t instance, uint16_t &cycles) const
+        return false;
+    }
+
+    char manufacturer_name[SMBUS_READ_BLOCK_MAXIMUM_TRANSFER+1];
```
Frontend should probably not know about SMBus.

But not a stopper for me, at least.

hendjoshsr71 replied:
Done. Moved this to backend.

peterbarker commented on libraries/AP_BattMonitor/AP_BattMonitor.cpp:
```diff
@@ -578,6 +616,71 @@ bool AP_BattMonitor::reset_remaining(uint16_t battery_mask, float percentage)
+void AP_BattMonitor::log_smart_battery_info() const
+{    
+    for(uint8_t i = 0; i < _num_instances; i++) {
+        if (get_type(i) > AP_BattMonitor::Type::ANALOG_VOLTAGE_AND_CURRENT) {
```
Urkh, no, please no.

Check the types explicitly, don't assume a range in an enumeration.

```suggestion
        if (get_type(i) == AP_BattMonitor::Type::ANALOG_VOLTAGE_AND_CURRENT) {
            continue;
        }
        if (get_type(i) == AP_BattMonitor::Type::BORING) {
            continue;
        }
```

peterbarker replied:
.... even better, just call `log_smart_battery_info()` on the backend.

Backends should log themselves....

hendjoshsr71 replied:
I noticed the SMBUS stuff, when moving logging to the new file, and am currently moving it to the backend of SMBUS with all of the piping in between. :)

hendjoshsr71 replied:
Moved to logging to backend.
Also explicitly checking the ENUM  now where I was doing the same thing in GCS_Common.

peterbarker commented on libraries/AP_BattMonitor/AP_BattMonitor.cpp:
```diff
@@ -578,6 +616,71 @@ bool AP_BattMonitor::reset_remaining(uint16_t battery_mask, float percentage)
+void AP_BattMonitor::log_smart_battery_info(uint8_t instance) const
+{
+    uint16_t cycles;
+    bool got_cycle_count = get_cycle_count(instance,cycles);
```
```suggestion
    const bool got_cycle_count = get_cycle_count(instance,cycles);
```

hendjoshsr71 replied:
fixed

peterbarker commented on libraries/AP_BattMonitor/AP_BattMonitor.cpp:
```diff
@@ -578,6 +616,71 @@ bool AP_BattMonitor::reset_remaining(uint16_t battery_mask, float percentage)
+    bool got_cycle_count = get_cycle_count(instance,cycles);
+
+    uint16_t design_capacity;
+    bool got_design_capacity = get_design_capacity(instance, design_capacity);
```
```suggestion
    const bool got_design_capacity = get_design_capacity(instance, design_capacity);
```

hendjoshsr71 replied:
fixed

peterbarker commented on libraries/AP_BattMonitor/AP_BattMonitor.cpp:
```diff
@@ -578,6 +616,71 @@ bool AP_BattMonitor::reset_remaining(uint16_t battery_mask, float percentage)
+
+    int32_t capacity_full = pack_capacity_mah(instance); //battery._params->_pack_capacity
+
+    if (capacity_full == 0) {
+        capacity_full = -1;
+    }
```
This breaks the pattern where you're using `got_foo` and substituting in some other value in `Write` below.  Choose one :-)

hendjoshsr71 replied:
Fixed to match the others. (Even though it adds an extra assignment?? ;) )

peterbarker commented on libraries/AP_BattMonitor/AP_BattMonitor_SMBus.h:
```diff
@@ -6,11 +6,13 @@
 #include <AP_HAL/I2CDevice.h>
 #include "AP_BattMonitor_Backend.h"
 #include <utility>
+#include <stdio.h>
```
Needed?

hendjoshsr71 replied:
Used, for the `snprintf's` in the serial number, product, device, and manufacturer name functions

hendjoshsr71 replied:
Moved to `AP_BattMonitor_SMBus.cpp` instead

peterbarker reviewed (changes requested):
Once this is merged we should send a message to the gcs-maintainers mailing list letting them know that they can request the message...

@WickedShell you're going to want to be happy with this before merge, I think?

hendjoshsr71 commented:
Thanks @peterbarker for the all the reviews!!

Moved the SMBUS stuff to the backends.
Moved the logging of the `BATI` message to the new `AP_BattMonito_Logging.cpp` file
Fixed hopefully all the trailing whitespaces...

peterbarker commented:
On Mon, 15 Mar 2021, Josh Henderson wrote:

> Used, for the snprintf's in the serial number, product, device, and manufacturer name functions

In the header?

hendjoshsr71 commented:
> On Mon, 15 Mar 2021, Josh Henderson wrote: Used, for the snprintf's in the serial number, product, device, and manufacturer name functions
> In the header?

Ahh I think I see your point it should be in the `AP_BattMonitor_SMBus.cpp' instead?

hendjoshsr71 commented:
Changed the capacity functions to int32_t as smbus spec of uint16 doesn't leave much room at 65,536 mAh
Still need to handle the strncpy truncation warnings
Could add manufactured date?

tridge commented on libraries/AP_BattMonitor/AP_BattMonitor_SMBus.h:
```diff
@@ -93,11 +126,17 @@ class AP_BattMonitor_SMBus : public AP_BattMonitor_Backend
     bool _has_cycle_count;          // true if cycle count has been retrieved from the battery
     bool _has_temperature;
+    uint8_t _device_name_len;
+    uint8_t _device_name[SMBUS_READ_BLOCK_MAXIMUM_TRANSFER+1]{};
```
don't init, wastes flash

hendjoshsr71 replied:
fixed

tridge commented on libraries/AP_BattMonitor/AP_BattMonitor_SMBus.h:
```diff
@@ -93,11 +126,17 @@ class AP_BattMonitor_SMBus : public AP_BattMonitor_Backend
+    uint8_t _device_name_len;
+    uint8_t _device_name[SMBUS_READ_BLOCK_MAXIMUM_TRANSFER+1]{};
+    uint8_t _manufacturer_name_len;
+    uint8_t _manufacturer_name[SMBUS_READ_BLOCK_MAXIMUM_TRANSFER+1]{};
```
don't init

hendjoshsr71 replied:
fixed

tridge commented on libraries/SITL/SIM_BattMonitor_SMBus.cpp:
```diff
@@ -10,19 +10,31 @@ SITL::SIM_BattMonitor_SMBus::SIM_BattMonitor_SMBus() :
-     // TODO: full capacity
-     set_register(SMBusBattDevReg::CYCLE_COUNT, (uint16_t(39U)));
+
+    set_register(SMBusBattDevReg::DESIGN_VOLTAGE, (uint16_t(50000U)));          // (mV) Design maximum voltage
```
indent wrong

hendjoshsr71 replied:
The indents here were originally 5 spaces. I fixed them to 4 spaces as a separate commit to match mine.

tridge commented on libraries/AP_BattMonitor/AP_BattMonitor_Logging.cpp:
```diff
@@ -45,3 +45,41 @@ void AP_BattMonitor_Backend::Log_Write_BCL(const uint8_t instance, const uint64_
+    char product_name[MAVLINK_MSG_SMART_BATTERY_INFO_FIELD_DEVICE_NAME_LEN]{}; 
+    get_product_name(product_name, ARRAY_SIZE(product_name));
+
+        struct log_BATI pkt{
```
indent wrong

hendjoshsr71 replied:
Fixed

tridge reviewed (changes requested):
i think we need to use HAL_MINIMIZE_FEATURES for this feature, to prevent the flash and ram cost for small boards with 1M flash

tridge commented on libraries/AP_BattMonitor/AP_BattMonitor_Logging.cpp:
```diff
@@ -45,3 +45,41 @@ void AP_BattMonitor_Backend::Log_Write_BCL(const uint8_t instance, const uint64_
+            cycles              : (uint16_t) (got_cycle_count ? cycles : UINT16_MAX),
+            serial_number       : {},
+            product_name        : {},
+            weight              : (double) NAN,
```
why have these fields?

hendjoshsr71 replied:
Removed.

peterbarker commented on libraries/AP_Logger/LoggerMessageWriter.cpp:
```diff
@@ -187,6 +187,8 @@ void LoggerMessageWriter_WriteSysInfo::reset()
 
 void LoggerMessageWriter_WriteSysInfo::process() {
     const AP_FWVersion &fwver = AP::fwversion();
+    const char *prot = hal.rcin->protocol();
+    const AP_BattMonitor &battery = AP::battery();
```
Use braces to give these a tighter scope for the relevant `case` clauses.

hendjoshsr71 replied:
Fixed

WickedShell commented on libraries/AP_BattMonitor/AP_BattMonitor_SMBus.h:
```diff
@@ -93,11 +126,17 @@ class AP_BattMonitor_SMBus : public AP_BattMonitor_Backend
 
     int32_t _serial_number = -1;    // battery serial number
-    uint16_t _full_charge_capacity; // full charge capacity, used to stash the value before setting the parameter
+    int32_t _full_charge_capacity; // full charge capacity, used to stash the value before setting the parameter
```
Why change this? It's a word off the SMBus, so it's only 16 bits. Increasing it here just uses more memory.

hendjoshsr71 replied:
Thanks! I had the frontend and backend capacity getters initially to uint16 but realized with only 65,536 mAh that they should likely be int32 but took that change too far to the SMBus.

WickedShell commented on libraries/AP_BattMonitor/AP_BattMonitor_SMBus.cpp:
```diff
@@ -30,6 +31,110 @@ bool AP_BattMonitor_SMBus::get_cycle_count(uint16_t &cycles) const
+bool AP_BattMonitor_SMBus::get_product_name(char *product_name, uint8_t buflen) const
+{
+
+    char manufacturer_name[SMBUS_READ_BLOCK_MAXIMUM_TRANSFER+1];
```
```suggestion
    char manufacturer_name[ARRAY_SIZE(_manufacturer_name)];
```

Don't rely upon remembering everywhere that has the length hard coded.

hendjoshsr71 replied:
Fixed. Good tip for the future!

hendjoshsr71 replied:
I've removed `_manufacturer_name` from the class. Thus this needed to used again.

WickedShell commented on libraries/AP_BattMonitor/AP_BattMonitor_SMBus.cpp:
```diff
@@ -30,6 +31,110 @@ bool AP_BattMonitor_SMBus::get_cycle_count(uint16_t &cycles) const
+bool AP_BattMonitor_SMBus::get_manufacturer_name(char * manufacturer_name, uint8_t buflen) const
+{    
+    if (_manufacturer_name_len) {
+        strncpy(manufacturer_name, (char*)_manufacturer_name, buflen -1);
```
You need to manually null terminate the array here, it could have been not initialized on the stack, which can lead to later reading of garbage memory. (This is the actual use case here). Also why would we bother copying more then _manufacturer_name_len of bytes? If we are storing the length why aren't we using it?

hendjoshsr71 replied:
Fixed it to copy the correct # of bytes and null terminate the last byte.

WickedShell commented on libraries/AP_BattMonitor/AP_BattMonitor_SMBus_Generic.cpp:
```diff
@@ -107,60 +106,37 @@ void AP_BattMonitor_SMBus_Generic::timer()
+    read();
+    
+    // check if PEC is supported
+    if (!check_pec_support()) {
```
PEC checks should always be done before any thing else. In fact until you've confirmed PEC either way you should not be attempting to read any other data from the battery. If we have error checking we need it to be used. That should be done before reading any other stats from the battery. There is a reason we always return if we haven't checked the PEC support in what was already here.

hendjoshsr71 replied:
Sorry about that bug.  I got it in the wrong position while trying to limit how often all of the system_info functions got called.

WickedShell commented on libraries/AP_BattMonitor/AP_BattMonitor_SMBus_Generic.cpp:
```diff
@@ -107,60 +106,37 @@ void AP_BattMonitor_SMBus_Generic::timer()
+    read_design_capacity();
+
+    // Set parameters with retrieved data
+    read();
```
You can't do this here, as this is on the timer thread. `read()` specifically is updating parameters, and issues MAVLink messages. Doing it here on the timer thread will lead to race conditions, and could crash the system.

hendjoshsr71 replied:
Fixed

hendjoshsr71 commented on libraries/AP_Logger/LoggerMessageWriter.cpp:
```diff
@@ -234,13 +236,17 @@ void LoggerMessageWriter_WriteSysInfo::process() {
+        FALLTHROUGH;
+
+    case Stage::BATT_CYCLES:
+        battery.log_smart_battery_info();
```
Fixed

WickedShell commented on libraries/AP_BattMonitor/AP_BattMonitor_SMBus.cpp:
```diff
@@ -20,49 +21,135 @@ void AP_BattMonitor_SMBus::init(void)
 bool AP_BattMonitor_SMBus::get_cycle_count(uint16_t &cycles) const
 {
-    if (!_has_cycle_count) {
+    WITH_SEMAPHORE(_dev->get_semaphore());
```
This is not a good idea. We should be reading this in the driver and stashing if we want it. The problem with exposing this is we start doing transactions from other threads, and since we configure the bus as SMBus this actually has a *very* long timeout on the transaction. If the battery (or something else) is very busy in the timer waiting on the hardware to catch up this could cause the main thread to stall while fetching the information for MAVLink for example.

hendjoshsr71 replied:
I understand your concerns and will revert those changes. 

The reason I was attempting this method was to limit storage and simplify functions required for rarely accessed info. Once the info. gets sent via mavlink or logged it no longer needs to be stored. I'm not sure if there is a way to accomplish that while limiting delays/timeouts to the battmon thread? 

WickedShell replied:
You can arbitrarily request a message from MAVLink though, and you might need to ensure someone who connects late gets it, which would argue for keeping it. The easiest way to minimize it is to store all the needed information in a struct you only allocated if any of the needed data was available.

hendjoshsr71 replied:
For the arbitrary mavlink request or late connection I was thinking of setting a flag to signal the battmon timer loop to handle building the message. (But I couldn't figure out how to do that. For logging it was obvious and I got that working in that manner.)

hendjoshsr71 replied:
set back to stashing

WickedShell commented on libraries/AP_BattMonitor/AP_BattMonitor_SMBus.cpp:
```diff
@@ -20,49 +21,135 @@ void AP_BattMonitor_SMBus::init(void)
 {
-    // nothing to be done here for actually interacting with the battery
-    // however we can use this to set any parameters that need to be set
+    WITH_SEMAPHORE(_dev->get_semaphore());
```
Same thing as above.

hendjoshsr71 replied:
set back to stashing

WickedShell commented on libraries/AP_BattMonitor/AP_BattMonitor_SMBus.cpp:
```diff
@@ -20,49 +21,135 @@ void AP_BattMonitor_SMBus::init(void)
+// returns true if the given register read of a character array was successful
+bool AP_BattMonitor_SMBus::get_name(AP_BattMonitor_SMBus::BATTMONITOR_SMBUS reg_name, char * name_out, uint8_t buflen) const
+{
+    WITH_SEMAPHORE(_dev->get_semaphore());
```
Same as above.

WickedShell reviewed (changes requested):
The changes to grab the device semaphore to try and do work from other threads actually looks extremely risky to me, and needs to be swapped back to stashing the information.

hendjoshsr71 commented:
Added `#if SMART_BATTMON_ENABLED` to get the build for PixHawk under 200 bytes, but was having trouble getting it lower than that.

hendjoshsr71 commented:
@WickedShell  Would you please review this when you get a chance?

WickedShell commented on libraries/AP_BattMonitor/AP_BattMonitor_Logging.cpp:
```diff
@@ -45,3 +49,106 @@ void AP_BattMonitor_Backend::Log_Write_BCL(const uint8_t instance, const uint64_
+    }
+#endif // HAL_NO_LOGGING
+
+#ifndef HAL_NO_GCS
```
Having a GCS with no logging won't work as all your private variables are inside the above no logging.

hendjoshsr71 replied:
Doh fixed!

WickedShell commented on libraries/AP_BattMonitor/AP_BattMonitor_Logging.cpp:
```diff
@@ -45,3 +49,106 @@ void AP_BattMonitor_Backend::Log_Write_BCL(const uint8_t instance, const uint64_
+    // get_cells_in_series only returns false if cell_count has not been determined yet
+    // if so set flag to call back logging or mavlink_send when cell_count is fixed
+    if (!got_cells_in_series) {
+        _instance = instance;
```
I'd go further and just split this into different functions. I know it add's some code (168 bytes in my test), but it's a lot clearer what's happening, and allows some cleanups, such as not fetching the data unless the GCS has space to send it, and removes the wrapper functions.

hendjoshsr71 replied:
Removed the instance stuff. See main comment about splitting the function

WickedShell commented on libraries/AP_BattMonitor/AP_BattMonitor_Backend.h:
```diff
@@ -71,11 +94,18 @@ class AP_BattMonitor_Backend
     AP_BattMonitor_Params               &_params;   // reference to this instances parameters (held in the front-end)
+    bool                                _log_smart_batt_info;       // set flag to log smart battery info message
+    bool                                _mavlink_smart_batt_info;   // set flag to send mavlink smart battery info message
+    uint8_t                             _instance;  // instance number
```
This already has a meaning and a role in the Sum backend, given that this directly assigns it sometimes, we should probably use a different name here (or cope without it, which would be my personal preference).

hendjoshsr71 replied:
Removed instance

WickedShell commented on libraries/AP_BattMonitor/AP_BattMonitor_SMBus_Generic.cpp:
```diff
@@ -107,60 +89,68 @@ void AP_BattMonitor_SMBus_Generic::timer()
-        return 0;
-    }
+            if (_mavlink_smart_batt_info) {
+                send_mavlink_smart_battery_info(_instance, _chan);
```
You can't send here. This is the I2C timer thread, and the GCS_MAVLink stuff isn't thread safe. You'd either have to set a flag for the GCS to poll at a regular interval, or just fail the first request if the data isn't available, and wait for a second one (which would be my personal preference, as it removes 2 members, and simplifies the GCS interaction, as well as fixing interactions when multiple links request the data at once).

hendjoshsr71 replied:
Removed this so that we now just notify the GCS of failure to gather all of the data. One day I'll learn which threads can interact with each other. :)  I assume that what I was doing with the logging thread was ok then?

WickedShell commented on libraries/AP_BattMonitor/AP_BattMonitor_Logging.cpp:
```diff
@@ -45,3 +49,106 @@ void AP_BattMonitor_Backend::Log_Write_BCL(const uint8_t instance, const uint64_
+            instance            : instance,
+            function            : MAV_BATTERY_FUNCTION_UNKNOWN,
+            type                : MAV_BATTERY_TYPE_UNKNOWN,
+            cells_in_series     : (uint8_t) (got_cells_in_series ? cells_in_series : 0),
```
You already have the cells in series if you made it here, no need for the ternary.

hendjoshsr71 replied:
Yup removed.

WickedShell commented on libraries/AP_BattMonitor/AP_BattMonitor_Logging.cpp:
```diff
@@ -45,3 +49,106 @@ void AP_BattMonitor_Backend::Log_Write_BCL(const uint8_t instance, const uint64_
+            UINT16_MAX,                     // (not provided) [mV] Minimum per-cell voltage when discharging. If not supplied set to UINT16_MAX value.
+            UINT16_MAX,                     // (not provided) [mV] Minimum per-cell voltage when charging. If not supplied set to UINT16_MAX value.
+            UINT16_MAX,                     // (not provided) [mV] Minimum per-cell voltage when resting. If not supplied set to UINT16_MAX value.
+            got_cells_in_series ? cells_in_series : 0,  // Number of battery cells in series. 0: field not provided.
```
You already have the cells in series if you made it here, no need for the ternary.

hendjoshsr71 replied:
Yup removed.

WickedShell commented on libraries/AP_BattMonitor/AP_BattMonitor_SMBus.cpp:
```diff
@@ -30,6 +31,74 @@ bool AP_BattMonitor_SMBus::get_cycle_count(uint16_t &cycles) const
 
+#if HAL_SMART_BATT_LOG_ENABLED
+// return true if serial number can be provided and fills in serial number argument
+bool AP_BattMonitor_SMBus::get_serial_number(char *serial_number, uint8_t buflen) const
```
I don't think a helper that does the snprintf is really worth it. The caller should probably handle this directly as needed, but I guess if you split it into 2 places then keeping this is maybe acceptable?

hendjoshsr71 replied:
It doesn't help at all. Getting rid of it saved 70ish-bytes

WickedShell commented on libraries/AP_BattMonitor/AP_BattMonitor_SMBus.h:
```diff
@@ -52,13 +57,36 @@ class AP_BattMonitor_SMBus : public AP_BattMonitor_Backend
+    bool get_full_charge_capacity(int32_t &full_charge_capacity) const override;
+
+    // returns true if the design voltage (maximum charging voltage) can be provided and fill it in
+    bool get_design_voltage(int32_t &design_voltage) const override;
```
These all need units. Design voltage should be a float in volts instead of integer millivolts, just to be consistent with the `voltage()` function. If you want to leave the design/charge capacity as mah (which is fine), then the names should end with `_mah`

hendjoshsr71 replied:
Added units in all the comments and made them volts and A-h. For the fields we are stashing in SMBus I kept the variables in mAh and mV and renamed them to be clear. 

For the design voltage in mV I would need to store it as float if it was in volts you decide?

WickedShell reviewed (changes requested):
We can't send MAVLink from the I2C timer thread, that needs to be cleaned up or we have bad race conditions.

hendjoshsr71 commented:
I'm removed the flags to retry logging or mavlink.  Thus we won't log or send mavlink until the 15 second cell check is complete. But will notify the GCS of the failure to send. This makes everything far simpler and properly adding the instance_num to constructors seems wasteful.

@WickedShell Your decision, but for now I've kept the the function that either sends mavlink or logs and renamed it to `exclusive_send_or_log_smart_battery_info`. Doing this saves 272B flash when I built for Durndal which seems significant from every call I've attended?

I think with the wrappers it isn't too convoluted. The most obtuse part of it is the #ifdefs. Which I could simplify by just using `#ifndef HAL_BUILD_AP_PERIPH` as that is were those are coming from?

BTW this fails the LoggerDocumentation test because it looks like that test doesn't handle the flag  `/` for A-h properly?

hendjoshsr71 commented on libraries/AP_BattMonitor/AP_BattMonitor_Logging.cpp:
```diff
@@ -45,3 +49,106 @@ void AP_BattMonitor_Backend::Log_Write_BCL(const uint8_t instance, const uint64_
+    send_smart_battery_info(instance, false, chan);
+}
+
+void AP_BattMonitor_Backend::send_smart_battery_info(const uint8_t instance, bool log_flag, const mavlink_channel_t chan)
```
I renamed this to be clear hopefully?

hendjoshsr71 commented on libraries/AP_BattMonitor/AP_BattMonitor.cpp:
```diff
@@ -582,6 +582,31 @@ bool AP_BattMonitor::reset_remaining_mask(uint16_t battery_mask, float percentag
+void AP_BattMonitor::send_mavlink_smart_battery_info(const uint8_t instance, const mavlink_channel_t chan) const
+{
+#if HAL_SMART_BATT_LOG_ENABLED
+    if (drivers[instance] != nullptr && get_type(instance) != Type::NONE) {
```
fixed

hendjoshsr71 commented on libraries/AP_BattMonitor/AP_BattMonitor.cpp:
```diff
@@ -582,6 +582,31 @@ bool AP_BattMonitor::reset_remaining_mask(uint16_t battery_mask, float percentag
+    AP_Logger *logger = AP_Logger::get_singleton();
+
+    for(uint8_t i = 0; i < _num_instances; i++) {
+        if (drivers[i] != nullptr && get_type(i) != Type::NONE) {
```
Fixed

hendjoshsr71 commented on libraries/AP_BattMonitor/AP_BattMonitor.cpp:
```diff
@@ -582,6 +582,31 @@ bool AP_BattMonitor::reset_remaining_mask(uint16_t battery_mask, float percentag
+
+    for(uint8_t i = 0; i < _num_instances; i++) {
+        if (drivers[i] != nullptr && get_type(i) != Type::NONE) {
+            if (logger != nullptr && logger->should_log(_log_battery_bit)) {
```
Done. BTW  See [here](https://github.com/hendjoshsr71/ardupilot/blob/ab5343f75436c5dc8e2b4a549074a6d1618de7b3/libraries/AP_BattMonitor/AP_BattMonitor.cpp#L303) for where the logging was moved to the backend. You may want that changed a bit?

hendjoshsr71 commented:
Rebased and updated mavlink reference to smartinfo message now in the mavlink upstream

tridge commented on libraries/AP_BattMonitor/AP_BattMonitor.cpp:
```diff
@@ -588,6 +588,37 @@ bool AP_BattMonitor::reset_remaining_mask(uint16_t battery_mask, float percentag
+#if HAL_SMART_BATT_LOG_ENABLED
+    AP_Logger *logger = AP_Logger::get_singleton();
+
+if (logger != nullptr && logger->should_log(_log_battery_bit)) {
```
indentation

hendjoshsr71 replied:
this got removed to solve the problem of logging information in case of a new log file being opened

tridge commented on libraries/AP_BattMonitor/LogStructure.h:
```diff
@@ -55,8 +56,41 @@ struct PACKED log_BCL {
+// @Field: ser: serial number in ASCII characters, 0 terminated. All 0: field not provided.
+// @Field: date: manufacture date (DD/MM/YYYY) in ASCII characters, 0 terminated. All 0: field not provided
+// @Field: name: product name encoded as 'device name'_'manufacturer'
+struct PACKED log_BATI {
```
should be a log write() call

tridge replied:
need to test that the log of the strings works

hendjoshsr71 replied:
switched to log_write and tested. 
![image](https://user-images.githubusercontent.com/69225461/121989943-f2488b00-cd6a-11eb-9efb-3bd711545861.png)

rmackay9 commented on libraries/GCS_MAVLink/GCS_Common.cpp:
```diff
@@ -4763,6 +4767,35 @@ void GCS_MAVLINK::send_set_position_target_global_int(uint8_t target_system, uin
+#if HAL_SMART_BATT_LOG_ENABLED
+    const AP_BattMonitor &battery = AP::battery();
+
+    for(uint8_t i = 0; i < battery.num_instances(); i++) {
```
nitpick: we normally have a space between "for" and the following bracket.

hendjoshsr71 replied:
fixed

rmackay9 commented on libraries/AP_BattMonitor/AP_BattMonitor_Logging.cpp:
```diff
@@ -45,3 +50,114 @@ void AP_BattMonitor_Backend::Log_Write_BCL(const uint8_t instance, const uint64_
+}
+
+// Exclusively either send or log the smart battery info. If log_flag true logs BATI message if false sends mavlink message
+bool AP_BattMonitor_Backend::exclusive_send_or_log_smart_battery_info(const uint8_t instance, bool log_flag, const mavlink_channel_t chan)
```
we don't normally have a single method that both sends status via mavlink and log to onboard logs. see "exclusive_send_or_log_smart_battery_info"

hendjoshsr71 replied:
@rmackay9 
Yeah this was my way to save flash space. I think it was 273 bytes in the code as it is currently. 

But it adds complexity doing it like this. I'm guessing clearer code is preferable?

hendjoshsr71 reviewed (commented):
Switched to log_write. It does work with the char[32] & char[64].

![image](https://user-images.githubusercontent.com/69225461/121989748-a7c70e80-cd6a-11eb-924b-e8a1ba07e33c.png)

Added a little bit more to the sitl drivers to test that it logged and sent messages properly with all three smbus types defined and an analog monitor type..

Added a check to stash the current log count and  log the message again. 
Added a flag to the battmonitor state that is set when the smart_info is ready from the backend driver to also log the info.

peterbarker commented on libraries/AP_BattMonitor/AP_BattMonitor_Logging.cpp:
```diff
@@ -70,3 +75,132 @@ void AP_BattMonitor_Backend::Log_Write_BCL(const uint8_t instance, const uint64_
+    get_product_name(smart_batt_info.product_name, ARRAY_SIZE(smart_batt_info.product_name));
+    get_manufacture_date(smart_batt_info.manufacture_date,  ARRAY_SIZE(smart_batt_info.manufacture_date));
+
+    const int32_t ser_num =  _params._serial_number;
```
Don't we want to cast this to `uint32_t`?  It is an artifact of our param system that we can't distinguish signed/unsigned well

hendjoshsr71 replied:
`_params._serial_number` is -1 until the smartbatt gets the serial number. And I need to check if that is the case to null terminate the string if no serial number was read.  Is this type of check ok `_params._serial_number != -1`?

peterbarker commented on libraries/AP_BattMonitor/AP_BattMonitor_Logging.cpp:
```diff
@@ -70,3 +75,132 @@ void AP_BattMonitor_Backend::Log_Write_BCL(const uint8_t instance, const uint64_
+        snprintf(smart_batt_info.serial_number, ARRAY_SIZE(smart_batt_info.serial_number), "%i", (signed) ser_num);
+    } else {
+        // ensure null termination when data is unavailable
+        if (0 < ARRAY_SIZE(smart_batt_info.serial_number)) {
```
This is a bit weird
```suggestion
        if (ARRAY_SIZE(smart_batt_info.serial_number) > 0) {
```

tridge commented:
added DelayMerge for delay after final rebase for 4.1.0

tridge commented on libraries/AP_BattMonitor/AP_BattMonitor_SMBus.cpp:
```diff
@@ -151,6 +311,52 @@ bool AP_BattMonitor_SMBus::read_word(uint8_t reg, uint16_t& data) const
+    }
+
+    // buffer to hold results (2 extra byte returned holding length and PEC)
+    const uint8_t read_size = bufflen + 1 + (_pec_supported ? 1 : 0);
```
looks like read_size needs to be 16 bit, otherwise is bufflen == 255 we wrap to zero and get stack corruption

peterbarker commented on libraries/GCS_MAVLink/GCS_Common.cpp:
```diff
@@ -4747,6 +4751,35 @@ void GCS_MAVLINK::send_set_position_target_global_int(uint8_t target_system, uin
+    }
+    return true;
+#else
+    return false;
```
This must return true.  These send messages can only return false in the case there's no buffer space, essentially.

You also need to test this PR with SMART_BATTERY_INFO disabled.

hendjoshsr71 reviewed (commented):
Fixed the GCS_send so that false only gets returned if there is no space on the link. 

Tested with 2 analog battmons, 1 smbus_generic, 1 rotoye, and 1 maxell. That logging and mavlink send works with my mavproxy submodule

Tested with SMART_BATTERY_INFO disabled and checked logs such that all makes sense for given monitor type.

amilcarlucas commented:
This requires 2440 bytes on Durandal (arducopter)

WickedShell commented on libraries/AP_BattMonitor/AP_BattMonitor_SMBus.h:
```diff
@@ -22,12 +22,17 @@ class AP_BattMonitor_SMBus : public AP_BattMonitor_Backend
-        BATTMONITOR_SMBUS_MANUFACTURE_DATA = 0x23,     // Manufacture Data
+        BATTMONITOR_SMBUS_DEVICE_NAME = 0x21,          // Device Name
+        BATTMONITOR_SMBUS_DEVICE_CHEMISTRY = 0x22,     // Battery chemistry type
+        BATTMONITOR_SMBUS_MANUFACTURE_DATA = 0x23      // Manufacture Data
```
Nitpick: leaving the comma doesn't break the compiler, and makes the diff smaller both for here, and future changes.

WickedShell commented on libraries/AP_BattMonitor/AP_BattMonitor_SMBus_Generic.cpp:
```diff
@@ -51,13 +49,14 @@ AP_BattMonitor_SMBus_Generic::AP_BattMonitor_SMBus_Generic(AP_BattMonitor &mon,
 {
-	// check if PEC is supported
-    if (!check_pec_support()) {
+    uint32_t tnow = AP_HAL::micros();
```
```suggestion
    const uint32_t tnow = AP_HAL::micros();
```

nitpick

WickedShell commented on libraries/AP_BattMonitor/AP_BattMonitor_SMBus_Generic.cpp:
```diff
@@ -118,60 +100,62 @@ void AP_BattMonitor_SMBus_Generic::timer()
-    const uint8_t read_size = bufflen + 1 + (_pec_supported ? 1 : 0);
-    uint8_t buff[read_size];
+    // check cell count
+    if (_state.healthy) {
```
Unless I've missed something healthy isn't actual true initially, and it's normally toggled by the polling loop. Why are we looking at it at all inside of the initialization block?

Especially since this is called repeatedly during polling until the initialized flag (which can only be set by entering here) doesn't this mean we might repeatedly poll for all the smart battery info for ever on a non healthy battery and massively bring down the I2C bus by accident?

WickedShell reviewed (changes requested):
Some nitpicks. I didn't have time to fully tell but the initialization bit still looks worrisome to me, as I'm not seeing a path where we don't try and spend a ton of I2C bandwidth talking to batteries that aren't there....

hendjoshsr71 reviewed (commented):
@WickedShell  Addressed most of your comments. But I have left a question or two about the termination condition for the initialization function.

hendjoshsr71 commented:
Rebased due to `#if HAL_GCS_ENABLED` change
